File: 93.pl
#!/usr/bin/env perl

# PODNAME: 93.pl
# ABSTRACT: Arithmetic expressions

## Author     : ian.sealy@sanger.ac.uk
## Maintainer : ian.sealy@sanger.ac.uk
## Created    : 2016-03-22

use warnings;
use strict;
use autodie;
use Getopt::Long;
use Pod::Usage;
use Carp;
use version; our $VERSION = qv('v0.1.0');

use Math::Combinatorics;

# Default options
my $max_digit = 9;    ## no critic (ProhibitMagicNumbers)
my ( $debug, $help, $man );

# Get and check command line options
get_and_check_options();

my @operators = (
    sub { my ( $n1, $n2 ) = @_; return $n1 + $n2 },
    sub { my ( $n1, $n2 ) = @_; return $n1 - $n2 },
    sub { my ( $n1, $n2 ) = @_; return $n1 * $n2 },
    sub { my ( $n1, $n2 ) = @_; return 0 if !$n2; return $n1 / $n2 },
);

my @op_perms;
## no critic (ProhibitMagicNumbers)
foreach my $i ( 0 .. 3 ) {
    foreach my $j ( 0 .. 3 ) {
        foreach my $k ( 0 .. 3 ) {
            ## use critic
            push @op_perms, [ $i, $j, $k ];
        }
    }
}
my @order_perms = permute( ( 1 .. 3 ) );

my $max_digits;
my $max_consec = 0;
## no critic (ProhibitMagicNumbers)
foreach my $digit_comb ( combine( 4, ( 1 .. $max_digit ) ) ) {
    ## use critic
    my @seen;
    foreach my $digit_perm ( permute( @{$digit_comb} ) ) {
        foreach my $op_perm (@op_perms) {
            foreach my $order_perm (@order_perms) {
                my @nums  = @{$digit_perm};
                my @ops   = @{$op_perm};
                my @order = @{$order_perm};
                while (@order) {
                    my $ordinal = shift @order;
                    my ( $num1, $num2 ) = splice @nums, $ordinal - 1, 2;
                    my ($op) = splice @ops, $ordinal - 1, 1;
                    splice @nums, $ordinal - 1, 0,
                      $operators[$op]->( $num1, $num2 );
                    @order = map { $_ > $ordinal ? $_ - 1 : $_ } @order;
                }
                my $num = shift @nums;
                next if int $num != $num;
                next if $num < 1;
                $seen[$num] = 1;
            }
        }
    }
    my $consec = 0;
    while ( $seen[ $consec + 1 ] ) {
        $consec++;
    }
    if ( $consec > $max_consec ) {
        $max_consec = $consec;
        $max_digits = join q{}, sort { $a <=> $b } @{$digit_comb};
    }
}

printf "%s\n", $max_digits;

# Get and check command line options
sub get_and_check_options {

    # Get options
    GetOptions(
        'max_digit=i' => \$max_digit,
        'debug'       => \$debug,
        'help'        => \$help,
        'man'         => \$man,
    ) or pod2usage(2);

    # Documentation
    if ($help) {
        pod2usage(1);
    }
    elsif ($man) {
        pod2usage( -verbose => 2 );
    }

    return;
}

__END__
=pod

=encoding UTF-8

=head1 NAME

93.pl

Arithmetic expressions

=head1 VERSION

version 0.1.0

=head1 DESCRIPTION

This script solves the Project Euler problem "Arithmetic expressions". The
problem is: Find the set of four distinct digits, a < b < c < d, for which the
longest set of consecutive positive integers, 1 to n, can be obtained, giving
your answer as a string: abcd.

=head1 EXAMPLES

    perl 93.pl --max_digit 4

=head1 USAGE

    93.pl
        [--max_digit INT]
        [--debug]
        [--help]
        [--man]

=head1 OPTIONS

=over 8

=item B<--max_digit INT>

The highest digit.

=item B<--debug>

Print debugging information.

=item B<--help>

Print a brief help message and exit.

=item B<--man>

Print this script's manual page and exit.

=back

=head1 DEPENDENCIES

None

=head1 AUTHOR

=over 4

=item *

Ian Sealy <ian.sealy@sanger.ac.uk>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2016 by Genome Research Ltd.

This is free software, licensed under:

  The GNU General Public License, Version 3, June 2007

=cut

File: 93.py
#!/usr/bin/env python

"""This script solves the Project Euler problem "Arithmetic expressions". The
problem is: Find the set of four distinct digits, a < b < c < d, for which the
longest set of consecutive positive integers, 1 to n, can be obtained, giving
your answer as a string: abcd.
"""

from __future__ import division
import argparse
from itertools import product, combinations, permutations


def main(args):
    """Arithmetic expressions"""

    operators = [
        lambda n1, n2: n1 + n2,
        lambda n1, n2: n1 - n2,
        lambda n1, n2: n1 * n2,
        lambda n1, n2: 0 if not n2 else n1 / n2
    ]

    op_perms = list(product(range(4), repeat=3))
    order_perms = list(permutations(range(1, 4)))

    max_digits = None
    max_consec = 0
    for digit_comb in combinations(range(1, args.max_digit + 1), 4):
        seen = [False] * (pow(9, 4) + 1)
        for digit_perm in permutations(digit_comb):
            for op_perm in op_perms:
                for order_perm in order_perms:
                    nums = list(digit_perm)
                    ops = list(op_perm)
                    order = list(order_perm)
                    while order:
                        ordinal = order.pop(0)
                        num1 = nums[ordinal-1:ordinal][0]
                        num2 = nums[ordinal:ordinal+1][0]
                        op = ops[ordinal-1:ordinal][0]
                        nums[ordinal-1:ordinal+1] = [operators[op](num1, num2)]
                        ops[ordinal-1:ordinal] = []
                        order = [i - 1 if i > ordinal else i for i in order]
                    num = nums[0]
                    if int(num) != num or num < 1:
                        continue
                    seen[int(num)] = True
        consec = 0
        while seen[consec + 1]:
            consec += 1
        if consec > max_consec:
            max_consec = consec
            max_digits = ''.join(str(i) for i in sorted(digit_comb))

    print(max_digits)

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Arithmetic expressions')
    parser.add_argument(
        'max_digit', metavar='MAX_DIGIT', type=int, default=9, nargs='?',
        help='The highest digit')
    args = parser.parse_args()

    main(args)
